### DISCUSSION TITLE
Session cookie removed on on every request.

### DISCUSSION THREAD

**Discussion by MilesStorm** [maxcountryman/axum-login#222] [Q&A] [answered]
Hello, i'm using axum-session with an sqlx session store, and I'm currently facing an issue where every time I've logged in a user. Any request they make to the authentication API will remove their session cookie.

![image](https://github.com/maxcountryman/axum-login/assets/29578565/c647f71b-798e-4683-af3d-5b5d60eecc04)

Full source can be found here: https://github.com/YousofMersal/website/tree/main/services/server

**Comment by maxcountryman** [comment]
How is the request being made? 

It sounds like the session cookie is not being included in the request based on the logs you shared.

**Reply by MilesStorm** [reply]
I've tried through a traditional browser that definetley recieved the cookie, and through an API debugging tool called postman. 

And i can log that a cookie exists, in the secondary route endpoint. 

![image](https://github.com/maxcountryman/axum-login/assets/29578565/f89b8376-31b3-4f7a-91ec-e752881504cf)

**Reply by maxcountryman** [reply]
Your session is empty so the cookie is removed.

**Reply by MilesStorm** [reply]
I'm not quite sure how to fix that, what does it mean that the session is empty? Does it not contain the user i logged in through axum-login? Looking at the database before it's deleted there's definitely a session there with an id and some data attached to it.

**Reply by maxcountryman** [reply]
I'm just reading the logs you shared; it shows the session is empty. I don't know why that's the case.

Can you try to reproduce what you're seeing with one of the examples in this repo?

**Reply by MilesStorm** [reply]
This should somewhat closely mimic the multi-auth exmaple, as i basically just copied that, and only modified the usage so that instead of an askama template, i repsond with JSON.

**Reply by MilesStorm** [reply]
Running the multi-auth example yields expected behaviour. So i'm really unsure on how to proceede. It should be really close to that example, yet it dosn't work.

**Reply by MilesStorm** [reply]
I figured out what the error was after trying one by one getting it closer to the multi arch example. In y original code i had 

```rust
let user = match auth_session
                .authenticate(Credentials::Password(creds.clone()))
                .await
            {
                Ok(Some(mut user)) => {
                    user.password = None;
                    user
                }
                Ok(None) => {
                    return (StatusCode::UNAUTHORIZED, "invalid password".to_string())
                        .into_response()
                }
                Err(_) => return StatusCode::INTERNAL_SERVER_ERROR.into_response(),
            };

            if auth_session.login(&user).await.is_err() {
                return StatusCode::INTERNAL_SERVER_ERROR.into_response();
            }
            
            // snip
``` 

But changing that to

```rust
            let mut user = match auth_session
                .authenticate(Credentials::Password(creds.clone()))
                .await
            {
                Ok(Some(user)) => user,
                Ok(None) => {
                    return (StatusCode::UNAUTHORIZED, "invalid password".to_string())
                        .into_response()
                }
                Err(_) => return StatusCode::INTERNAL_SERVER_ERROR.into_response(),
            };

            if auth_session.login(&user).await.is_err() {
                return StatusCode::INTERNAL_SERVER_ERROR.into_response();
            }

            user.password = None;
            // snip
```

I wasn't aware that axum-login used the password field in it's hash, which i guess makes the hash of the user not match and causes it to delete the session on the next request?

**Reply by maxcountryman** [reply]
If the hash doesn't match it will invalidate the login state and force the client to authenticate again. So that seems plausible. 

The auth hash doesn't have to be a password hash but it should be unique per user and the idea is that it will invalidate when a user changes important state (like their password).

**Reply by MilesStorm** [reply]
I though that the hash would be the user id as that should be unique, but i can't quite figure out where the hash is set, so i still don't quite understand how the error happened, i just know it it got fixed now. Where in the multi-auth example is it set where the auth-hash is made of?

**Reply by maxcountryman** [reply, marked as answer]
It's up to your implementation of AuthUser. But I would be careful setting it a user ID: while it is unique, it won't account for invalidating sessions when a user changes their passwords (that's okay if you don't want that but it's generally a good practice).

**Reply by MilesStorm** [reply]
Oh of course! Thank you now i understand how it works!

```rust
impl AuthUser for User {
    type Id = i64;

    fn id(&self) -> Self::Id {
        self.id
    }

    fn session_auth_hash(&self) -> &[u8] {
        if let Some(access_token) = &self.access_token {
            return access_token.as_bytes();
        }

        if let Some(password) = &self.password {
            return password.as_bytes();
        }

        &[]
    }
}
```

Now it all makes perfect sense, thank you for your time and explanation!
